### PR TITLE
gogs: init at v0.9.97

### PR DESCRIPTION
**Pull request by schneefux**
I updated #13566 to the latest release and improved the code a bit.

Tested:
- [x] repository mirroring
- [x] SSH authentication
- [x] `.tar.gz` downloads
- [x] PAM _disabled_
- [x] MySQL/Postgres setup
- [ ] fresh install

cc @valeriangalliat @kamilchm @yorickvP

### CODE REVIEW

File: pkgs/applications/version-management/gogs/default.nix [hunk 1]
```diff
@@ -0,0 +1,54 @@
+{ stdenv, buildGoPackage, fetchFromGitHub, makeWrapper
+, git, bash, openssh, pam
+, sqliteSupport ? true
+#, pamSupport ? false
+}:
+
+buildGoPackage rec {
+  name = "gogs-${version}";
+  version = "v0.9.97";
```
Comment (primeos) [inline pkgs/applications/version-management/gogs/default.nix]: > The version part of the name attribute must start with a digit (following a dash) — e.g., "hello-0.3.1rc2". [nixpkgs manual - 10.2. Package naming](https://nixos.org/nixpkgs/manual/#sec-package-naming)

I made the same mistake a few days ago...

File: pkgs/applications/version-management/gogs/default.nix [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+  src = fetchFromGitHub {
+    owner = "gogits";
+    repo = "gogs";
+    rev = "${version}";
```
Comment (primeos) [inline pkgs/applications/version-management/gogs/default.nix]: You could use `rev = "v${version}";` :wink:

**Review by primeos — CHANGES_REQUESTED** [review]

File: pkgs/applications/version-management/gogs/default.nix (hunk 1 above)
Comment (schneefux) [inline pkgs/applications/version-management/gogs/default.nix]: @primeos @groxxda `v` is removed.

File: pkgs/applications/version-management/gogs/deps.nix [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+    goPackagePath = "github.com/lib/pq";
+    fetch = {
+      type = "git";
+      url = "https://github.com/lib/pq";
+      rev = "50761b0867bd1d9d069276790bcd4a3bccf2324a";
+      sha256 = "1bz46slzri958wdnvlvyx12bp3pmy8iny5m0m16i3n8h1q84dhp5";
+    };
+  }
+  {
+    goPackagePath = "github.com/mattn/go-sqlite3";
```
Comment (Mic92) [inline pkgs/applications/version-management/gogs/deps.nix]: this probably needs the sqlite package

File: pkgs/applications/version-management/gogs/default.nix [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  src = fetchFromGitHub {
+    owner = "gogits";
+    repo = "gogs";
+    rev = "v${version}";
+    sha256 = "151mmd8h5zd4bvafd42nsky0m9gblf5fcpd6jacqcrry1796hxk9";
+  };
+
+  patchPhase = ''
+    substituteInPlace models/repo.go \
+      --replace '#!/usr/bin/env %s\n' "#!${bash}/bin/bash\\n#%s\\n"
```
Comment (LnL7) [inline pkgs/applications/version-management/gogs/default.nix]: This doesn't look right, you probably want to use `${coreutils}/bin/env` here

File: pkgs/applications/version-management/gogs/default.nix [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  ''; # replace shebang and ignore first argument (SCRIPT_TYPE in configuration) in TPL_UPDATE_HOOK
+
+  buildInputs = [ makeWrapper ]
+  #  ++ stdenv.lib.optional pamSupport pam
+  ;
+
+  buildFlags = "-tags"
```
Comment (LnL7) [inline pkgs/applications/version-management/gogs/default.nix]: `buildFlags` should be a list if you specify multiple arguments, then you don't have to worry about the space prefix.

File: pkgs/applications/version-management/gogs/deps.nix (hunk 3 above)
Comment (schneefux) [inline pkgs/applications/version-management/gogs/deps.nix]: It works fine without.
Other packages ([1](https://github.com/NixOS/nixpkgs/blob/98a9d815e05dd56aba0f9040d96092335cd90444/pkgs/shells/elvish/default.nix), [2](https://github.com/NixOS/nixpkgs/blob/98a9d815e05dd56aba0f9040d96092335cd90444/pkgs/development/tools/textql/default.nix)) that depend on `go-sqlite3` also don't include it.

File: pkgs/applications/version-management/gogs/default.nix (hunk 4 above)
Comment (schneefux) [inline pkgs/applications/version-management/gogs/default.nix]: Thanks!

**Comment by schneefux** [comment]
After updating and garbage collection, pushing fails with an error like `remote: hooks/update: line 3: /nix/store/53skjfhn19djc5mljp6vhr5mjm1gznnj-go1.7-gogs-v0.9.97-bin/bin/gogs: No such file or directory`
As stated in [Gogs' troubleshooting section](https://gogs.io/docs/intro/troubleshooting.html) the admin needs to rewrite all hooks via the admin panel when the Gogs binary path changes. How to deal with this or how do you notify the user?

**Comment by Mic92** [comment]
@schneefux this will happen after every upgrade, right?

**Comment by schneefux** [comment]
@Mic92 Yes/No. After an update, the hooks will use the old Gogs version until the user does `nix-collect-garbage`, and then it will fail.

**Comment by Mic92** [comment]
@schneefux could be those hooks fixed in the module on service restart?

**Comment by schneefux** [comment]
@Mic92 not as far as I know. I just opened an issue [here](https://github.com/gogits/gogs/issues/3796).

**Comment by Mic92** [comment]
@schneefux the alternative would be sed....

**Comment by schneefux** [comment]
@Mic92 
> the alternative would be sed regexp.

What do you think about
```bash
HOOKS=$(find ${gogs.repositoryRoot} -mindepth 4 -maxdepth 4 -type f -wholename "*git/hooks/*")
sed -i 's,"/nix/store/[a-z0-9.-]+/bin/gogs,${gogs.bin}/bin/gogs,g' $HOOKS
sed -i 's,"/nix/store/[a-z0-9.-]+/bin/env,${coreutils.bin}/bin/env,g' $HOOKS
sed -i 's,"/nix/store/[a-z0-9.-]+/bin/perl,${perl.bin}/bin/perl,g' $HOOKS
```
?

**Comment by Mic92** [comment]
@schneefux by default only the `update` hook containing gogs should by affected, right?

**Comment by schneefux** [comment]
> by default only the update hook containing gogs should by affected, right?

Yes, but I think it's better to update all hooks - for example if a user copies the `post-update.sample` hook.
I added the sed expressions to the module's preStart: https://github.com/NixOS/nixpkgs/pull/19363

**Comment by rnhmjoj** [comment]
What's the status on this? Is there something needed to be done? I'd love to see this merged.

**Comment by schneefux** [comment]
Updated to latest release
It's been running fine for months on my server, together with the module (#19363).

**Comment by LnL7** [comment]
I lost track of this, I'll test this after my other stuff is done building.

**Review by LnL7 — APPROVED** [review]

**Comment by rnhmjoj** [comment]
Thanks you!

**Pull request merged**